kratos/solving_strategies/schemes/residual_based_bdf_scheme.h

```diff
@@ -0,0 +1,582 @@
+ * with for order 2 (BDF2):
+ *  -# \f$ c_0 = \frac{1.5}{dt} \f$ 
+ *  -# \f$ c_1 = \frac{-2.0}{dt} \f$ 
+ *  -# \f$ c_2 = \frac{0.5}{dt} \f$  
+ * 
+ * The LHS and RHS can be defined as:
+ *      \f[ RHS = \mathbf{f}_{ext} - \mathbf{M} \frac{d(\dot{u}_{n0})}{dt} - \mathbf{D} \frac{d(u_{n0})}{dt} - \mathbf{K} u_{n0} \f]
+ * and 
+ *      \f[ LHS = \frac{d(-RHS)}{d(u_{n0})} = c_0^2 \mathbf{M} + c_0 \mathbf{D} + K \f]
+ * @note This implies that elements are expected to be written in terms 
+ * of a variable with two time derivatives
+ * <a href="https://mediatum.ub.tum.de/doc/1223319/80942.pdf">Main reference</a> 
+ * @todo Create a BibTeX file https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdcite
+ * @author Vicente Mataix Ferrandiz
+ */
+template<class TSparseSpace,  class TDenseSpace>
+class ResidualBasedBDFScheme
+    : public ResidualBasedImplicitTimeScheme<TSparseSpace, TDenseSpace>
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    KRATOS_CLASS_POINTER_DEFINITION( ResidualBasedBDFScheme );
+
+    typedef Scheme<TSparseSpace,TDenseSpace>                                  BaseType;
+    
+    typedef ResidualBasedImplicitTimeScheme<TSparseSpace,TDenseSpace> ImplicitBaseType;
+
+    typedef typename ImplicitBaseType::TDataType                             TDataType;
+
+    typedef typename ImplicitBaseType::DofsArrayType                     DofsArrayType;
+
+    typedef typename Element::DofsVectorType                            DofsVectorType;
+
+    typedef typename ImplicitBaseType::TSystemMatrixType             TSystemMatrixType;
+
+    typedef typename ImplicitBaseType::TSystemVectorType             TSystemVectorType;
+
+    typedef typename ImplicitBaseType::LocalSystemVectorType     LocalSystemVectorType;
+
+    typedef typename ImplicitBaseType::LocalSystemMatrixType     LocalSystemMatrixType;
+
+    typedef ModelPart::NodesContainerType                               NodesArrayType;
+
+    typedef ModelPart::ElementsContainerType                         ElementsArrayType;
+
+    typedef ModelPart::ConditionsContainerType                     ConditionsArrayType;
+
+    typedef typename BaseType::Pointer                                 BaseTypePointer;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /**
+     * Constructor.
+     * The BDF method
+     * @param Order The integration order
+     * @todo The ideal would be to use directly the dof or the variable itself to identify the type of variable and is derivatives
+     */
+    ResidualBasedBDFScheme(const std::size_t Order = 2)
+        :ImplicitBaseType(),
+         mOrder(Order)
+    {
+        // Allocate auxiliary memory
+        const std::size_t num_threads = OpenMPUtils::GetNumThreads();
+        
+        mVector.dotun0.resize(num_threads);
+        mVector.dot2un0.resize(num_threads);
+        
+        // Doing a minimal check
+        KRATOS_ERROR_IF(mOrder < 1) << "ERROR:: Not possible to compute a BDF of order 1" << std::endl;
```

Comment:
shouldnt  it be `ERROR:: Not possible to compute a BDF of order smaller than 1`?

Reply:
...true